raft/src/main/java/org/apache/kafka/raft/QuorumState.java

```diff
@@ -34,28 +34,29 @@
 public class QuorumState {
     public final int localId;
     private final Logger log;
-    private final ElectionStore store;
+    private final QuorumStateStore store;
     private final Set<Integer> voters;
     private EpochState state;
 
     public QuorumState(int localId,
                        Set<Integer> voters,
-                       ElectionStore store,
+                       QuorumStateStore store,
                        LogContext logContext) {
         this.localId = localId;
         this.voters = new HashSet<>(voters);
         this.store = store;
         this.log = logContext.logger(QuorumState.class);
     }
 
-    public void initialize(long endOffset) throws IOException {
+    public void initialize(OffsetAndEpoch offsetAndEpoch) throws IOException {
         // We initialize in whatever state we were in on shutdown. If we were a leader
         // or candidate, probably an election was held, but we will find out about it
         // when we send Vote or BeginEpoch requests.
+        store.initialize(offsetAndEpoch);
```

Comment:
The point I was trying to make before is that validation of the initialized position and exception handling can be done here so that it is valid for all implementations (including the mock). What I had in mind is something like the following:
```java
try {
  ElectionState election = store.readElectionState();
  if (election.epoch < offsetAndEpoch.epoch) {
    log.warn("Epoch from quorum-state file is smaller than last written epoch in the log");
    becomeUnattachedFollower(offsetAndEpoch.epoch);
  } 
} catch (Exception e) {
  log.warn("Failed to read quorum-state file, blah blah blah", e);
  becomeUnattachedFollower(offsetAndEpoch.epoch);
}
```
Then we don't need `ElectionStore.initialize`.

Reply:
One issue with this approach is that the readElectionState won't throw any exception in this case, as we did the loading of the quorum state when we construct the store. Maybe we could just add the `if (election.epoch < offsetAndEpoch.epoch)` validation here without a try-catch, since if exception was thrown from `becomeUnattachedFollower` then it should fatal for a write IO.